lib/Serialization/DocFormat.h

```diff
@@ -0,0 +1,85 @@
+//===--- DocFormat.h - The internals of swiftdoc files ----------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2018 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file Contains various constants and helper types to deal with serialized
+/// documentation info (swiftdoc files).
+///
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_SERIALIZATION_DOCFORMAT_H
+#define SWIFT_SERIALIZATION_DOCFORMAT_H
+
+#include "llvm/Bitcode/RecordLayout.h"
+
+namespace swift {
+namespace serialization {
+
+using llvm::BCArray;
+using llvm::BCBlob;
+using llvm::BCFixed;
+using llvm::BCGenericRecordLayout;
+using llvm::BCRecordLayout;
+using llvm::BCVBR;
+
+/// Magic number for serialized documentation files.
+const unsigned char SWIFTDOC_SIGNATURE[] = { 0xE2, 0x9C, 0xA8, 0x07 };
+
+/// Serialized swiftdoc format major version number.
+///
+/// Increment this value when making a backwards-incompatible change, which
+/// should be rare. When incrementing this value, reset SWIFTDOC_VERSION_MINOR
+/// to 0.
+const uint16_t SWIFTDOC_VERSION_MAJOR = 1;
+
+/// Serialized swiftdoc format minor version number.
+///
+/// Increment this value when making a backwards-compatible change that might
+/// be interesting to test for. However, if old swiftdoc files are fully
+/// compatible with the new change, you do not need to increment this.
+///
+/// To ensure that two separate changes don't silently get merged into one
+/// in source control, you should also update the comment to briefly
+/// describe what change you made. The content of this comment isn't important;
+/// it just ensures a conflict if two people change the module format.
+/// Don't worry about adhering to the 80-column limit for this line.
+const uint16_t SWIFTDOC_VERSION_MINOR = 1; // Last change: skipping 0 for testing purposes
+
+/// The record types within the comment block.
+///
+/// Be very careful when changing this block; it must remain stable. Adding new
+/// records is okay---they will be ignored---but modifying existing ones must be
+/// done carefully. You may need to update the version when you do so.
```

Comment:
What is the effect of adding a new 'field' at the end of an existing record, could you clarify it in the comments ?
Can we determine how many fields a record has at deserialization so we don't have to update and check for a version number ?

Reply:
You can kind of do that—see https://github.com/apple/swift/blob/bae3964c07a976ee0978c0c2400e23e3252e60e8/lib/Serialization/ModuleFile.cpp#L200-L216—but BCRecordLayout doesn't support it. You have to read records manually in that case. Adding new records isn't really that expensive, though, so I don't think it's too much of a problem to do that.

For a stable format, the version number is supposed to be higher-level than just "did I change something". If it makes no difference, it's fine not to update it (and if you really don't want to use it, we could just drop the "minor" version entirely).

Reply:
I think it's good that you added the minor version, as an extra thing we may find useful to check,  I'm mainly trying to understand if we should update it when adding a new field.

Adding a new field would force us to switch to fully manually reading the record ? Would we be able to do a check and switch to using a different BCRecordLayout to deserialize that record, and could this check occur without needing to update the minor version ?

Pointing in the documentation to the example that you gave on how to manually check for the number of fields would be useful.

Reply:
It's unfortunately worse than that: adding a new field will **break existing tools**, because BCRecordLayout currently asserts that the number of fields matches exactly. https://github.com/apple/swift-llvm/blob/stable/include/llvm/Bitcode/RecordLayout.h#L261

We could change that, of course, or add a new API `readIgnoringTrailingFields` or something, but the point is that you can only add new fields if there aren't existing tools expecting a fixed number of fields.